src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3899,4 +3816,163 @@ bool ViewProviderSketch::isInEditMode() const
 {
     return editCoinManager != nullptr;
 }
+void ViewProviderSketch::generateContextMenu() {
+    int selectedEdges = 0;
+    int selectedLines = 0;
+    int selectedConics = 0;
+    int selectedPoints = 0;
+    int selectedConstraints = 0;
+
+    Gui::MenuItem menu;
+    menu.setCommand("Sketcher context");
+
+    std::vector<Gui::SelectionObject> selection = Gui::Selection().getSelectionEx(0, Sketcher::SketchObject::getClassTypeId());
+
+    // if something is selected, count different elements in the current selection
+    if (selection.size() > 0) {
+        const std::vector<std::string> SubNames = selection[0].getSubNames();
+
+        for (auto& name : SubNames) {
+            if (name.substr(0, 4) == "Edge") {
+                ++selectedEdges;
+
+                int geoId = std::atoi(name.substr(4, 4000).c_str()) - 1;
+                if (geoId >= 0) {
+                    const Part::Geometry* geo = getSketchObject()->getGeometry(geoId);
+                    if (isLineSegment(*geo)) {
+                        ++selectedLines;
+                    }
+                    else {
+                        ++selectedConics;
+                    }
+                }
+            }
+            else if (name.substr(0, 4) == "Vert") {
+                ++selectedPoints;
+            }
+            else if (name.substr(0, 4) == "Cons") {
+                ++selectedConstraints;
+            }
+        }
+        // build context menu items depending on the selection
+        if (selectedEdges >= 1 && selectedPoints == 0) {
+            menu << "Sketcher_Dimension";
+            if (selectedConics == 0) {
+                menu << "Sketcher_ConstrainHorVer"
+                << "Sketcher_ConstrainVertical"
+                << "Sketcher_ConstrainHorizontal"
+                << "Sketcher_ConstrainBlock";
+            }
+            else if (selectedConics > 1 && selectedLines == 0){
+                menu << "Sketcher_ConstrainCoincidentUnified"
+                    << "Sketcher_ConstrainTangent"
+                    << "Sketcher_ConstrainEqual";
+            }
+            else if (selectedConics == 1 && selectedLines == 1){
+                menu << "Sketcher_ConstrainTangent";
+            }
+            if (selectedLines > 1 && selectedConics == 0) {
+                menu << "Sketcher_ConstrainParallel";
+                    if (selectedLines ==2) {
```

Comment:
There are still few formatting mishaps - you can try to change your settings of vscode to enable reformatting on save:

https://github.com/microsoft/vscode/issues/44075#issuecomment-678231273

Reply:
oh it worked. the whole file was in //canlang-format off